### ISSUE TITLE
Circuit should ignore gaia

### ISSUE THREAD

**Issue by GoogleFrog** [rlcevg/CircuitAI#59]
Gaia units should not be processed by circuit at all. In particular it shouldn't avoid or attempt to attack them.

**Comment by GoogleFrog** [comment]
I was reminded that zombies use gaia. Could circuit instead read a unitRulesParam to check whether it is supposed to completely ignore a unit? The issue is that circuit is trying to destroy planetwars structures in missions that exist to add flavour.

**Comment by rlcevg** [comment]
Yes, just name the unitRulesParam and give it proper permissions: enemy/AI should be able to read it even without direct los, just radar, otherwise AI will hunt such radar dots.

**Comment by GoogleFrog** [comment]
Ok, how about "ignoredByCircuitAI"?

**Comment by GoogleFrog** [comment]
https://github.com/ZeroK-RTS/Zero-K/commit/7c0b46195a8f31e9ec7d8bd1d52ba2e797100e25

**Comment by rlcevg** [comment]
"ignoredByAI" or "ignored_by_ai" (i don't remember what's the prefered style), no need to tie it with circuit keyword.

**Comment by ashdnazg** [comment]
maybe use "SetNeutral"?

**Comment by GoogleFrog** [comment]
Sure, I'll change it to "ignoredByAI".

I dislike "SetNeutral" because it risks forgetting what the parameter is for. Also with Spring.SetUnitNeutral(unitID, true) it looks redundant.

**Comment by ashdnazg** [comment]
I meant using SetUnitNeutral, not a param

**Comment by rlcevg** [comment]
AI can read if unit is neutral (Unit::IsNeutral), but i don't know what that is (didn't use it and din't see it in other AIs). But if it's supposed to work then i can check for Unit::IsNeutral and ignore such units (though i don't know its visibility params... checking engine code...)

**Comment by rlcevg** [comment]
Nope, https://github.com/spring/spring/blob/develop/rts/ExternalAI/AICallback.cpp#L709, AI won't see if unit is neutral if it's not in los, and will hunt radar dots, i don't think it's best solution.

**Comment by ashdnazg** [comment]
can AI get params if unit is not in los?

**Comment by rlcevg** [comment]
it can get unit: https://github.com/spring/spring/blob/develop/rts/ExternalAI/SSkirmishAICallbackImpl.cpp#L3655
and should see params if it has proper mask: [LuaRulesParams::RULESPARAMLOS_PUBLIC_MASK](https://github.com/spring/spring/blob/develop/rts/ExternalAI/SSkirmishAICallbackImpl.cpp#L125)

But i didn't try in practice of course :)

**Comment by ashdnazg** [comment]
The neutral not being visible with radar is a discrepancy with lua and should be fixed in engine.

**Comment by rlcevg** [comment]
i'll make PR on pair with #53 (another los issues) if it won't fix itself

**Comment by rlcevg** [comment]
https://github.com/rlcevg/CircuitAI/commit/4713adfe014f8a4652f5d69e60d9a6cee787f78f
After this commit AI should ignore IsNeutral or ignoredByAI enemies (it doesn't track changes, only 1st encounter matters).

**Comment by GoogleFrog** [comment]
It is more powerful if circuit doesn't ignore neutral units.

**Comment by rlcevg** [comment]
Ok, will remove IsNeutral from enemy exclusion (though i thought it's exactly the purpose of this param).

**Comment by rlcevg** [comment]
https://github.com/rlcevg/CircuitAI/commit/ce173e3ea505e20934710c3e43d00d08620c6749

**Comment by ashdnazg** [comment]
I don't follow @GoogleFrog either

**Comment by GoogleFrog** [comment]
If a unit is neutral it is because the game made it neutral. If it is intended to be ignored by circuit then the game can set the required unitRulesParam. I can't think of a time when a neutral unit would not want to be ignored by circuit but such a thing may occur and if it occurs I would rather not wait on the long build cycle of circuit.

In short, games have more control if circuit doesn't automatically ignore neutral units.

**Comment by ashdnazg** [comment]
Making a unit neutral means that other teams' units shouldn't fire on it. That also takes into account automatic commands.
Having a neutral unit that an AI needs to attack should _never_ happen.

**Comment by GoogleFrog** [comment]
This ticket is about ignoring the unit entirely. Perhaps not attacking it is smart but it needs to do some other behaviour. I have now thought up an example.

Say there is some ability that makes a an army invulnerable for 30 seconds. If a unit is invulnerable then other units should not automatically attack it and the simplest way of doing this is via Spring.SetUnitNeutral. If circuit completely ignores the army then it will do very poorly, it may run smaller forces into the army unaware that they will die. If it doesn't ignore the army then it will at least avoid it with most of its forces and probably retreat when it takes too many losses.

**Comment by ashdnazg** [comment]
> and the simplest way of doing this is via Spring.SetUnitNeutral

Can't say I agree, this unit is not neutral, it shouldn't be marked as such. There are other ways to prevent auto-attacking and/or priorities.

Though obviously, if you're misusing Spring.SetUnitNeutral, then it can't be used by any AI for any purpose whatsoever.
